control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -3743,17 +3744,66 @@ func (r *HostedControlPlaneReconciler) reconcileOperatorLifecycleManager(ctx con
+
+			if r.ManagementClusterCapabilities.Has(capabilities.CapabilityImageStream) {
+				catalogsImageStream := manifests.CatalogsImageStream(hcp.Namespace)
+				if !overrideImages {
+					if _, err := createOrUpdate(ctx, r, catalogsImageStream, func() error {
+						return olm.ReconcileCatalogsImageStream(catalogsImageStream, p.OwnerRef, isImageRegistryOverrides)
+					}); err != nil {
+						return fmt.Errorf("failed to reconcile catalogs image stream: %w", err)
+					}
+				} else {
+					if _, err := util.DeleteIfNeeded(ctx, r, catalogsImageStream); err != nil {
+						return fmt.Errorf("failed to remove OLM Catalog ImageStream: %w", err)
+					}
 				}
-			} else if r.ManagementClusterCapabilities.Has(capabilities.CapabilityImageStream) {
-				if _, err := util.DeleteIfNeeded(ctx, r, catalogsImageStream); err != nil {
-					return fmt.Errorf("failed to remove OLM Catalog ImageStream: %w", err)
+			} else {
+				for name, catalog := range olm.CatalogToImage {
+					pullSecret := common.PullSecret(hcp.Namespace)
+					if err := r.Get(ctx, client.ObjectKeyFromObject(pullSecret), pullSecret); err != nil {
+						return fmt.Errorf("failed to get pull secret for namespace %s: %w", hcp.Namespace, err)
+					}
+
+					parts := strings.Split(catalog, ":")
+					if len(parts) != 2 {
+						return fmt.Errorf("invalid catalog format %s, expected 'registry:tag'", catalog)
+					}
+					catalogImage, tag := parts[0], parts[1]
+
+					catalogImageParts := strings.Split(catalogImage, "/")
+					catalogRepo := catalogImageParts[len(catalogImageParts)-1]
+					catalogRegistry := strings.Join(catalogImageParts[:len(catalogImageParts)-1], "/")
+
+					if isImageRegistryOverrides != nil {
+						if mirror, exists := isImageRegistryOverrides[catalogRegistry]; exists && len(mirror) > 0 {
+							catalogRegistry = mirror[0]
+						}
+					}
+
+					catalogImageWithRegistryOverride := fmt.Sprintf("%s/%s:%s", catalogRegistry, catalogRepo, tag)
+					listDigest, err := registryclient.GetListDigest(ctx, catalogImageWithRegistryOverride, pullSecret.Data[corev1.DockerConfigJsonKey])
+					if err != nil {
+						return fmt.Errorf("failed to get manifest for image %s: %v", catalogImageWithRegistryOverride, err)
+					}
+
+					imageOverride := fmt.Sprintf("%s/%s@%s", catalogRegistry, catalogRepo, listDigest.String())
+
+					catalogOverrides := map[string]*string{
+						"redhat-operators":    &p.RedHatOperatorsCatalogImageOverride,
+						"certified-operators": &p.CertifiedOperatorsCatalogImageOverride,
+						"community-operators": &p.CommunityOperatorsCatalogImageOverride,
+						"redhat-marketplace":  &p.RedHatMarketplaceCatalogImageOverride,
+					}
+
+					if override, exists := catalogOverrides[name]; exists {
+						if ptr.Deref(override, "") == "" {
+							*override = imageOverride
+						}
+					} else {
+						return fmt.Errorf("unknown catalog %s", name)
```

Comment:
Could you make this a little more verbose for debuggin purposes in the future? Something like "unknown catalog for catalog image override".